### PR TITLE
Document `vshard.router.map_callrw()`

### LINKED ISSUE #2015: [5pt] vshard.router.map_callrw()
# Map-reduce in vshard: vshard.router.map_callrw()

**Product:** Tarantool, vshard (external module)
**Since:** 0.1.17 (vshard version; vshard is compatible with Tarantool versions >= 1.9)
**Audience/target:** developers; Tarantool implementation teams
**Root document:** https://www.tarantool.io/en/doc/latest/reference/reference_rock/vshard/vshard_index/
https://www.tarantool.io/en/doc/latest/reference/reference_rock/vshard/vshard_api/#vshard-api-reference-router-public-api

**SME:** @ Gerold103 
**Peer reviewer:** @

# Details

`vshard.router.map_callrw()` implements consistent map-reduce over
the entire cluster. Consistency means all the data was accessible,
and didn't move during map requests execution.

It is useful when need to access potentially all the data in the
cluster or simply huge number of buckets scattered over the
instances and whose individual `vshard.router.call()` would take
too long.

`Map_callrw()` takes name of the function to call on the storages,
arguments in the format of array, and not required options map.
The only supported option for now is timeout which is applied to
the entire call. Not to individual calls for each storage.
```
vshard.router.map_callrw(func_name, args[, {timeout = <seconds>}])
```

The chosen function is called on the master node of each
replicaset with the given arguments.

In case of success `vshard.router.map_callrw()` returns a map with
replicaset UUIDs as keys and results of the user's function as
values, like this:
```
{uuid1 = {res1}, uuid2 = {res2}, ...}
```
If the function returned `nil` or `box.NULL` from one of the
storages, it won't be present in the result map.

In case of fail it returns nil, error object, and optional
replicaset UUID where the error happened. UUID may not be returned
if the error wasn't about a concrete replicaset.

For instance, the method fails if not all buckets were found even
if all replicasets were scanned successfully.

Handling the result looks like this:
```Lua
res, err, uuid = vshard.router.map_callrw(...)
if not res then
    -- Error.
    -- 'err' - error object. 'uuid' - optional UUID of replicaset
    -- where the error happened.
    ...
else
    -- Success.
    for uuid, value in pairs(res) do
        ...
    end
end
```

Map-Reduce in vshard works in 3 stages: Ref, Map, Reduce. Ref is
an internal stage which is supposed to ensure data consistency
during user's function execution on all nodes.

Reduce is not performed by vshard. It is what user's code does
with results of `map_callrw()`.

Consistency, as it is defined for map-reduce, is not compatible
with rebalancing. Because any bucket move would make the sender
and receiver nodes 'inconsistent' - it is not possible to call a
function on them which could simply access all the data without
doing `vshard.storage.bucket_ref()`.

This makes Ref stage very intricate as it must work together with
rebalancer to ensure neither of them block each other.

For this storage has a scheduler specifically for bucket moves and
storage refs which shares storage time between them fairly.

Definition of fairness depends on how long and frequent the moves
and refs are. This can be configured using storage options
`sched_move_quota` and `sched_ref_quota`. See more details about
them in the corresponding doc section.

The scheduler configuration may affect map-reduce requests if they
are used a lot during rebalancing.

Keep in mind that it is not a good idea to use too big timeouts
for `map_callrw()`. Because the router will try to block the
bucket moves for the given timeout on all storages. And in case
something will go wrong, the block will remain for the entire
timeout. This means, in particular, having the timeout longer
than, say, minutes is a super bad way to go unless it is for
tests only.

Also it is important to remember that `map_callrw()` does not
work on replicas. It works only on masters. This makes it unusable
if at least one replicaset has its master node down.

*   Related development issues and/or commits: https://github.com/tarantool/vshard/commit/a394e3f4008a7436f011c81028dcaadc270eebf6

# Definition of done

- [x] Create a new page with the description of map-reduce implementation in `vshard`
- [x] Restructure [API Reference page](https://www.tarantool.io/en/doc/latest/reference/reference_rock/vshard/vshard_api/)

### PR DESCRIPTION
**Pull request by xuniq**
Fixes #2015 

Deploy (API Reference): https://develop.tarantool.io/en/doc/gh-2015-map-reduce/reference/reference_rock/vshard/vshard_api/

Deploy (map_callrw()): https://develop.tarantool.io/en/doc/gh-2015-map-reduce/reference/reference_rock/vshard/vshard-router/#lua-function.vshard.router.map_callrw

### CODE REVIEW

File: doc/reference/reference_rock/vshard/vshard-router.rst [hunk 1]
```diff
@@ -0,0 +1,872 @@
...
+            else
+                -- Success.
+                for uuid, value in pairs(res) do
+                    ...
+                end
+            end
+
+    Map-Reduce in vshard can be divided into three stages: Ref, Map, and Reduce.
+    The first stage ensures data consistency while executing the user's function on all nodes.
+    Consistency, defined for map-reduce, is not compatible with rebalancing.
```
Comment (veod32) [inline doc/reference/reference_rock/vshard/vshard-router.rst:367]: The idea is not clear. Try to re-write.

**Review by veod32 — CHANGES_REQUESTED** [review]

File: doc/reference/reference_rock/vshard/vshard-router.rst [hunk 2]
```diff
@@ -0,0 +1,872 @@
...
+    Any bucket move would make the sender and receiver nodes 'inconsistent',
+    so it is impossible to call a function on them to access all the data
+    without :ref:`vshard.storage.bucket_ref() <storage_api-bucket_ref>`.
+    It makes the Ref stage intricate, as it should work together with the rebalancer to ensure
+    they do not block each other.
+
+    For this, the storage has a special scheduler for bucket moves and storage refs.
```
[the remark this answers is not in the document]
Comment (Gerold103) [reply doc/reference/reference_rock/vshard/vshard-router.rst:374]: Storage ref is a volatile counter defined on each instance. It is incremented when a map-reduce request comes and decremented when it ends. It is almost the same as bucket ref, but it pins the entire instance with all its buckets, not just a single bucket.

**Review by Gerold103 — COMMENTED** [review]
Thanks for the patch!

File: doc/reference/reference_rock/vshard/vshard-router.rst [hunk 3]
```diff
@@ -0,0 +1,872 @@
...
+    It shares storage time between them fairly.
+    The definition of fairness depends on how long and frequent the moves and refs are.
+    It can be configured using storage options ``sched_move_quota`` and ``sched_ref_quota``.
```
[the remark this answers is not in the document]
Comment (xuniq) [reply doc/reference/reference_rock/vshard/vshard-router.rst:386]: https://github.com/tarantool/doc/issues/2014

File: doc/reference/reference_rock/vshard/vshard-router.rst [hunk 4]
```diff
@@ -0,0 +1,886 @@
...
+    they do not block each other.
+
+    For this, the storage has a special scheduler for bucket moves and storage refs.
+    Storage ref is a volatile counter defined on each instance.
+    It is incremented when a map-reduce request comes and decremented when it ends.
+    Storage ref pins the entire instance with all its buckets, not just a single bucket (like bucket ref).
+
+    The scheduler shares storage time between bucket moves and storage refs fairly.
+    The distribution depends on how long and frequent the moves and refs are.
+    It can be configured using the storage options ``sched_move_quota`` and ``sched_ref_quota``.
```
Comment (veod32) [inline doc/reference/reference_rock/vshard/vshard-router.rst:387]: Make :ref: links to the description of these options

**Review by veod32 — APPROVED** [review]

**Pull request merged**